oeps/oep-0045/decisions/0001-tutor-as-replacement-for-edx-configuration.rst

```diff
@@ -0,0 +1,328 @@
+<https://docs.tutor.overhang.io/configuration.html#custom-open-edx-docker-image>`__.  The latter is how one adds custom
+themes, xblocks, or even uses `a fork of edx-platform
+<https://docs.tutor.overhang.io/configuration.html#running-a-fork-of-edx-platform>`__.  In operational terms it is no
+harder than how one would accomplish the same with ``edx/configuration``.
+
+For many things that Tutor doesn't explicitly support (and for which there is no desire/need to include in the core), it
+is possible to `write plugins <https://docs.tutor.overhang.io/plugins/gettingstarted.html>`__.  With this engine, one
+can add additional services and configure the rest of the platform to use them. One can imagine writing plugins that
+allow deployment of the different Open edX MFEs, for instance.
+
+A note on theming: it would be ideal if it were possible to store static assets separately so that a single base image
+could be used for multiple instances, but that is not currently the case.
+
+
+Criterion 6. Migration from the native installation
+...................................................
+
+There is no easily found documentation on migrating to a Tutor-managed installation from a native one, aside from `this
+discussion in the Overhang.io forums <https://discuss.overhang.io/t/migrating-from-bitnami-to-tutor/31>`__.
+
+Assuming the source installation is running the same release as the target one, it would be a reasonably straightforward
+matter of launching a new Tutor stack (with all the desired customization applied, of course), dumping the state of the
+databases, then importing them.  Nevertheless, the exact steps need to be performed, tested, and written down.
+
+
+Criterion 7. Project health
+...........................
+
+Tutor as an open source project is `3 and a half years old
+<https://github.com/overhangio/tutor/commit/bdd1a41f6207275bd24c98136fc567af2ea48b7f>`__.  It saw a peak of activity `in
+2019 <https://github.com/overhangio/tutor/graphs/code-frequency>`__, and has since then maintained a steady pace of
+improvement. The company behind it, `Overhang.IO <https://overhang.io/>`__, is declaredly committed to open source
+values and offers `Tutor deployment <https://overhang.io/tutor/>`__ as a service, and thus have a financial incentive to
+continue maintaining it.
+
+It is, however, still effectively maintained `by a single developer
+<https://github.com/overhangio/tutor/graphs/contributors>`__: Regis Behmo. This poses a significant risk: if for some
+unforeseen reason Regis can no longer maintain Tutor, the Open edX community would again be orphaned. Thus, it is
+suggested that some form of co-governance/co-maintenance be instituted. Regis has `tentatively agreed
+<https://discuss.openedx.org/t/lets-talk-about-the-native-installation/3269/5>`__.  Hopefully, with more people and
+companies using Tutor, more contributors will be willing to take on this responsibility.
```

Comment:
Tutor is open to external contributions, and Tutor users [already](https://github.com/overhangio/tutor/pulls?q=is%3Apr+is%3Aclosed) submit PRs. This PRs are usually reviewed and merged in a timely manner, so I am not quite sure we need to amend this process just yet.

I want to keep full control over what goes in the Tutor core app as well as the supported plugins, so I will remain a bottleneck for PR reviewing and merging. However I must be more transparent when it comes to making important architectural changes. This can take the form of pre-release announcements, ADRs, synchronous conversations or forum topics. For instance, the recent changes introduced in Koa (uwsgi, Caddy, Redis...) would have been discussed prior to release.

My position when it comes to making changes to the core of Tutor is the following: can you implement your changes in a plugin? If yes, you should develop such a plugin and it is your responsibility to maintain it. If not, what changes must we make in Tutor to allow you to implement such a plugin? If those changes are not possible or desirable, then, and only then should you fork Tutor. 

Of course, that last scenario is not desirable, and in my experience it occurs very rarely.

Let me phrase the question differently: what organizational changes would the community like to see implemented to gain more confidence in the future of Tutor?

Reply:
This seems like the crux of the matter:
> I want to keep full control over what goes in the Tutor core app as well as the supported plugins, so I will remain a bottleneck for PR reviewing and merging.

We need more specifics about what is the Tutor core, what are the supported plugins, and what is the community installation.  We don't want work on the community installation to be gated by one reviewer, no matter how responsive.

Reply:
@nedbat, agreed.  I'm going to suggest adding this topic (and any pending points in this ADR) to the next BTR working group meeting agenda for the 18th.  That way we can spend a few minutes figuring out the way forward.  @regisb, is that ok?

Reply:
> We don't want work on the community installation to be gated by one reviewer, no matter how responsive.

Let me play devil's advocate here: why don't we want the community installation to be gated by one company? (notice I say "company", not "reviewer"). The only scenarios I see where that might be an issue are when:

- The community installation is suddenly broken: this never happened, because the only external dependencies are docker.io and the tutor packages on pypi.
- A security patch is published and needs to be added urgently to the base Docker image: such issues should (must) be mitigated by being warned in advance of the publication of a security patch.

Am I missing something?

> I'm going to suggest adding this topic (and any pending points in this ADR) to the next BTR working group meeting agenda for the 18th. That way we can spend a few minutes figuring out the way forward. @regisb, is that ok?

Sure @arbrandes; you are very much welcome to add items on the BTR meetup agenda. I think that to get a productive conversation you should come with a set of clear expectations. For instance: "we need PRs to be reviewed within X hours" or "we need a backup strategy in case you stop working on it".

Reply:
@regisb 

> Let me play devil's advocate here: why don't we want the community installation to be gated by one company?

Because we just finally managed to at least partially stop having to rely on one company, gating all the PRs on the rest of the Open edX project, with the introduction of core committers. The switch to Tutor is supposed to be a flagship example of  how working together as a community can be more beneficial - it would be quite the irony if this ended up adding a new piece to the project who doesn't even have that same level of shared responsibility and decision-making... Can we show how going further this way can work well? It's definitely a painful and frightening step to take - it's not for nothing that it took a long time for edX to start going in that direction -- but we know better already, no?

Reply:
I think that it's very difficult to compare the state of Tutor and Open edX, as open source projects, both in terms of: code base (3k vs 800k LOC), reactivity to PRs/issues (hours vs weeks) and core team size (1 vs 100). In Open edX it was decided to pick non-edX core committers to address community inclusion. But I do not think that the same problem (and thus the same solution) applies to Tutor.

But then, I'm not even sure having core committers is what you are asking for, because you did not answer my above question:

> what organizational changes would the community like to see implemented to gain more confidence in the future of Tutor?

Reply:
@regisb Yes, Tutor is more like one of the projects or teams of Open edX, rather than something to compare to Open edX as a whole in terms of size. But that's part of my point - we are proposing that Tutor becomes the official installation method for the Open edX project - so that would make it part of the Open edX project (and a critical part of it, that is!), and even if it is owned by you rather than by an edX team, it is still important to follow the project best practices. It's like if OpenCraft said that we wouldn't accept any other core committers on blockstore, despite it being one of the core components of the platform...

> what organizational changes would the community like to see implemented to gain more confidence in the future of Tutor?

I can't speak for the whole community, but on my side I think it should at least be part of the core committer program yes, and allow other Open edX core committers (which you would pick, based on demonstrated skills & matching philosophy) to have merge rights, as well as the ability to merge PRs independently from you. Though that's a minimum, imho it could be good to use the opportunity to push the concept further, and this could be a discussion - I would be interested in hearing what you would suggest.

Reply:
> we are proposing that Tutor becomes the official installation method for the Open edX project - so that would make it part of the Open edX project (and a critical part of it, that is!), and even if it is owned by you rather than by an edX team, it is still important to follow the project best practices.

The core committers program is a band-aid patch for a deeper issue, which is that non-edX PRs take too long (if ever) to review. The CC program only partially resolves this issue: many PRs, even [trivial](https://github.com/edx/edx-platform/pull/25771) ones, take months before they get merged. I would be more than glad to trade my CC badge for a more reactive project stewardship.

Core committers are a best practice in the context of Open edX, but not in the rest of the open source world. There are [plenty](https://github.com/redis/redis) of [great](https://github.com/elastic/elasticsearch) [projects](https://github.com/WordPress/WordPress) out there that strive under the umbrella of a single organization.

Tutor works because at some point I decided to do things differently than the rest of Open edX. I will not rely on a suboptimal governance model simply because it's the default in Open edX.

> It's like if OpenCraft said that we wouldn't accept any other core committers on blockstore, despite it being one of the core components of the platform...

Who is responsible for maintaining Blockstore? If you are saying that OpenCraft is taking on this responsibility, then I'll be perfectly fine with the project having no external core contributor. IMHO it would actually be a big win for the whole community.

When all you have is a hammer, all problems look like a nail. Now, why do I think that joining the core contributors program is the wrong hammer for that nail? There are several reasons:

1. It moves some knowledge away from the creator and maintainer of the project (aka: me), thus making it more difficult to quickly debug and resolve issues.
2. It dilutes the project responsibility: when multiple people are responsible for a task, no one is.
3. It reduces the visibility of my company: at the risk of stating the obvious, I am building a business around Tutor, and it is important that this business remains viable, otherwise the project will wither. This will only work if Overhang.IO is widely considered as the creator and manager of the project.

I should mention that there are alternatives to joining the core committers program:

- Core _reviewers_ who review PRs but do not have merge rights
- Product owners who help test features and establish a roadmap
- Plugin maintainers who extend the core software outside the umbrella of the original company
- Support engineers who answer questions on the forums

If Tutor becomes the new native installation (as I hope it does), maybe I'll become swamped in PRs, discussion topics and Github issues. I doubt it, as I am dealing with a small 3k LOC code base which I wrote pretty much all by myself, but if it's the case then I'll be happy to bring more people onboard to help. But today, it's just too soon to commit to this.

> it should at least be part of the core committer program yes, and allow other Open edX core committers (which you would pick, based on demonstrated skills & matching philosophy) to have merge rights, as well as the ability to merge PRs independently from you. Though that's a minimum, imho it could be good to use the opportunity to push the concept further, and this could be a discussion - I would be interested in hearing what you would suggest.

Here's what I can commit to: _if_ reactivity to PRs becomes a problem at some point, _if_ there are external contributors with a proven track-record of feature development on Tutor, _then_ joining the core committers program will be _one_ of the solutions I'll consider.

Does this sound reasonable to you?

Reply:
@regisb,

> The core committers program is a band-aid patch for a deeper issue, which is that non-edX PRs take too long (if ever) to review.

While the CC program is indeed an attempt to remediate PR reviews, I would hesitate to call it a band-aid.  I see it more as an acknowledgement that projects such as the Linux Kernel, OpenStack, and Debian (listed in order of increasing community stewardship), get it right.

I know that comparing Tutor, a 3 Kloc project, to OpenStack, at 12 Mloc with thousands of contributors, is hardly fair.  But Open edX itself makes more sense, so bear with me for a while.  In _my_ ideal world, Open edX would be entirely community driven, just like OpenStack, and financed by the same community via indirect methods such as donations or an organization membership.  There would be a whole new class of problems, but some of the ones we see today could be elegantly solved: among them, the time-to-review we're discussing here, and also, for instance, the Release Notes issue we're trying to handle elsewhere.

This is, alas, not the case, nor is it likely to ever come to pass.  But, and this is the crux: it shouldn't stop us from picking elements from such projects that make sense to our particular zeitgeist.  _Specially_ if the parent organization is willing to hear us out as a community.

And here's the snag: while Tutor is far from suffering from the problems the CC program was created to solve - indeed, it looks like it was created precisely as a way to solve a part of them! - it doesn't send the right message if the community asks for it from edX, but gives Overhang a pass.

I could put forward several practical advantages for everybody, including Overhang, in having such "lieutenants" doing volunteer work in Tutor (speed, agility, resilience, etc).  But that is not why I brought this up in the ADR: as @antoviaque says, whatever replaces edx/configuration is critical.  We would love to have it give off the same "community signals" we're seeing from edX.

Note that I'm not suggesting that either edX or Overhang give up control of Open edX and Tutor, respectively.  And just for completeness, I *definitely* don't think either should be forked, *ever*.  Like the projects you mention, there are many successful codebases under enlightened company leadership and thriving communities.  I happen to think edX is one of them, and Overhang even more so.  It's just that having such positions as core committers is, above all, a signal of mutual trust: both that the community trusts Overhang, and that Overhang trusts the community.

Finally, as Xavier said, you'd be the one to pick committers, if any live up to your standards.  You'd be free to "fire" them and revert their changes, if they somehow break your trust.  I don't see the current Open edX core committers going against the direction set by edX, and I wouldn't expect anything different from Tutor "core committers" - if those are what they end up being called.

Reply:
We are running around in circles. I am saying that adding core committers would be detrimental to the project, and you are saying that not having core committers would be sending you a bad signal. In order to move the needle, I drafted a proposal for a maintainers program which actually makes sense in the context of Tutor: https://github.com/overhangio/tutor/pull/395/ Please let me know what you think.

Reply:
Thank you for the proposal @regisb - I'll review and post a few comments about it calmly next week, but it's really useful to know what would work for you. I agree that we have a better chance to get to a consensus by starting from what we each want and are willing to do, rather than the other way around. Cheers! :)

Reply:
@regisb I've done a pass of review on https://github.com/overhangio/tutor/pull/395 and left a few comments there,